test/tutorials/lambda_theory.jl

```diff
@@ -0,0 +1,209 @@
+
+# This can be done via a `LambdaAnalysis` datastructure which we can include in an
+# `EClass`. We overload Egraphs.make such that whenever we add a new enode to
+# the egraph we keep track of the free variables.
+
+const LambdaAnalysis = Set{Symbol}
+
+getdata(eclass) = isnothing(eclass.data) ? LambdaAnalysis() : eclass.data
+
+function EGraphs.make(g::EGraph{ExprType,LambdaAnalysis}, n::VecExpr) where ExprType
+  v_isexpr(n) || LambdaAnalysis()
+  if v_iscall(n)
+    h = v_head(n)
+    op = get_constant(g, h)
+    args = v_children(n)
+    eclass = g[args[1]]
+    free = getdata(eclass)
+
+    if op == Variable
+      push!(free, get_constant(g, v_head(eclass.nodes[1])))
+
+    elseif op == Let
+      v, a, b = args[1:3]
+      adata = getdata(g[a])
+      bdata = getdata(g[b])
+      union!(free, adata)
+      delete!(free, v)
+      union!(free, bdata)
+
+    elseif op == λ
+      v, b = args[1:2]
+      bdata = getdata(g[b])
+      union!(free, bdata)
+      delete!(free, v)
+
+    elseif op == Apply
+      l, v = args[1:2]
+      ldata = getdata(g[l])
+      vdata = getdata(g[v])
+      union!(free, ldata)
+      union!(free, vdata)
+
+    end
+    return free
+  end
+end
+
+EGraphs.join(from::LambdaAnalysis, to::LambdaAnalysis) = union(from,to)
+
+function fresh_var_generator()
+  idx = 0
+  function generate()
+    idx += 1
+    chars = collect(string(idx))
+    subs = map(digit -> Char(Int(digit) + Int('₀') - Int('0')), chars)
+    Symbol("a$(String(subs))")
+  end
+end
+
+freshvar = fresh_var_generator()
+
+# The final ruleset then looks like below and correctly renames variables when needed:
+
+λT = @theory v e c v1 v2 a b body begin
+  Let(v, e, c::Any) --> c
+  Let(v1, e, Variable(v1)) --> e
+  Let(v1, e, Variable(v2)) => v1 == v2 ? e : Variable(v2)
+  Let(v1, e, λ(v1, body)) --> λ(v1, body)
+  Apply(λ(v, body), e) --> Let(v, e, body)
+  Let(v, e, Apply(a, b)) --> Apply(Let(v, e, a), Let(v, e, b))
+  Let(v1, e, λ(v2, body)) => if isfree(_egraph,e,v2)
+    fresh = freshvar()
+    λ(fresh, Let(v1, e, Let(v2, Variable(fresh), body)))
+  else
+    λ(v2, Let(v1, e, body))
+  end
+end
+
+x = Variable(:x)
+y = Variable(:y)
+ex = Apply(λ(:x, λ(:y, Apply(x,y))), y)
+g = EGraph{LambdaExpr,LambdaAnalysis}(ex)
+saturate!(g, λT)
+@test λ(:a₄, Apply(y, Variable(:a₄))) == extract!(g, astsize)
+
+
+# With the above we can implement, for example, Church numerals.
+
+s = Variable(:s)
+z = Variable(:z)
+n = Variable(:n)
+zero = λ(:s, λ(:z, z))
+one = λ(:s, λ(:z, Apply(s,z)))
+two = λ(:s, λ(:z, Apply(s,Apply(s,z))))
+suc = λ(:n, λ(:x, λ(:y, Apply(x, Apply(Apply(n,x),y)))))
+
+# Compute the successor of `one`:
+
+freshvar = fresh_var_generator()
+g = EGraph{LambdaExpr,LambdaAnalysis}(Apply(suc, one))
+params = SaturationParams(
+  timeout = 20,
+  scheduler = Schedulers.BackoffScheduler,
+  schedulerparams = (6000, 5),
+  timer = false,
+)
+saturate!(g, λT, params)
+extract!(g, astsize)
+
+# which is the same as `two` up to $\alpha$-conversion:
+
+two
```

Comment:
Can you make this a test please?

Reply:
done!:) I initially didnt, because its a bit awkward because of the necessary alpha conversion